source/opt/eliminate_dead_code_pass.cpp

```diff
@@ -0,0 +1,129 @@
+// Copyright (c) 2016 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "eliminate_dead_code_pass.h"
+
+#include <algorithm>
+#include <unordered_set>
+
+#include "def_use_manager.h"
+#include "instruction.h"
+#include "log.h"
+#include "operand.h"
+#include "reflect.h"
+
+namespace spvtools {
+namespace opt {
+
+namespace {
+
+// Returns true if the given |inst| is a dead instruction:
+// * generating result id, and
+// * no uses or uses are all debug instructions, and
+// * if is variable-defining instruction, having internal visibility.
+bool IsDeadInst(const analysis::DefUseManager& def_use_mgr,
+                const ir::Instruction& inst) {
+  // Instructions not generating result ids are used for their side effects.
+  if (inst.result_id() == 0) return false;
```

Comment:
I don't think this is quite correct. Consider the various atomics which generate a result id, but have side effects which must be performed even if the result is ignored.